File: Project_and_Analysis_of_Algorithms/graph_search_applications/c++/bipartite.cpp
#include "graphs.h"

bool isBipartite(adjacencyMap(string) graph) {
    visitedMap(string) visited = initVisited(graph);
    map<string, bool> color = visited;
    
    // Choses a node for each component.
    for(auto const&[node, neighbours] : graph) {
        if (visited[node]) continue;

        // Starts a search queue for a specific node. 
        queue<string> q; q.push(node);

        // Attribute color 1 to the node,
        color[node] = true;
        visited[node] = true;
        while(!q.empty()) {
            string curNode = q.front(); q.pop();
            for (string neighbour : graph[curNode]) {
                if (visited[neighbour] && color[curNode] == color[neighbour]) {
                    // Found an odd cycle (two neighbours with same color).
                    return false;
                } else if(!visited[neighbour]) {
                    // Inverts the color of the cur node based on the neighbour color.
                    color[neighbour] = !(color[curNode]);
                    visited[neighbour] = true;
                    q.push(neighbour);
                }
            }
        }
    }

    // No odd cycle was found.
    return true;
}

int main(void) {
    cout << isBipartite(G2) << "\n";
    return EXIT_SUCCESS;
}

File: Project_and_Analysis_of_Algorithms/graph_search_applications/python/bipartite.py
from graphs import G2
from collections import deque

def bipartite(graph: dict[str, list[str]]) -> dict[str, bool] | bool:
    """Bipartite test using BFS."""
    visited: dict[str, bool] = {node: False for node in graph.keys()}
    color: dict[str, bool] = visited.copy()
    for node in graph.keys():
        queue: deque[str] = deque([node])
        while (len(queue) > 0):
            cur_node: str = queue.popleft()
            for neighbour in graph[cur_node]:
                if not visited[neighbour]:
                    queue.append(neighbour)
                    visited[neighbour] = True
                    color[neighbour] = not(color[cur_node])
                elif color[cur_node] == color[neighbour]:
                    return False

    return color


def main():
    print(bipartite(graph=G2))
    return


if __name__ == "__main__":
    main()
